#{set title:'T-RMSD Tutorial' /}

<style type="text/css">
div.box img {
  -moz-box-shadow: 0px 5px 10px #909090 ;
  -webkit-box-shadow: 0px 5px 10px #909090 ;
  box-shadow: 0px 5px 10px #909090 ;
  
  border: 1px solid #666;
}

.box ul { padding-left: 1.3em; padding-bottom: 1em }
.box ul li { padding-bottom: .5em; list-style: disc;  }
.box ul li em { padding-right: .2em }
</style>

<h1>T-RMSD Tutorial
<small>A quick start to the T-RMSD protein structural classification</small>
</h1>

<div class="box">
<h2><span>Description</span></h2>

<p>
The T-RMSD web server allows a fine-grained structural classification of proteins using the 
T-RMSD method (Tree based on Root Mean Square Deviation), see <a href='#ref' target='_blank'>[1, 2]</a>.
</p>
<p>
Providing a set of proteins sequences, it produces a structural tree with support 
values for each cluster node, reminiscent to bootstrap values.

These values, associated with the tree topology, allow a quantitative estimate of structural distances
between proteins or group of proteins defined by the tree topology. 
</p>

</div>

<div class="box">
<h2><span>Data input</span></h2>

<ul >
<li>
Enter the sequences in the input field (marked with '1' in the picture below) or alternatively 
provide them using the <i>Upload link</i> (2).<br>
The protein sequences have to be in a standard format, either <i>FASTA</i> or <i>ClustalW</i>. 
</li>

<li>You may access to T-RMSD advanced settings by clicking the link <i>Show more options</i> (3). </li>

<li>Provide your email address (4) if you want to be notified when the job complete.</li>

<li>Press the button <i>submit</i> (5) to submit your request.</li>

</ul>

<p>
Please note that the computation can take some minutes to be fulfilled. You don't need to keep your browser open to wait 
for the result. You can close it or navigate away to another page. 
</p>
<p>Use the <i>History</i> link, in top-bar menu to get access to the computation result at any moment.</p>

<p> <img alt="T-RMSD input form" src="public/images/trmsd1.png" width="728" height="612"> </p>

</div>




<div class="box" >
<h2><span>Output interpretation</span></h2>

<p>The result page contains the following sections: </p>

<ul>
<li>
<em>MSA</em>
the MSA, colored according to the T-RMSD scheme (it is not a T-Coffee consistency based color scheme). 
The T-RMSD color scheme is related to the informativeness of each ungapped column, 
from blue (poorly discriminant) to red (highly discriminant). 
</li>

<li><em>Structural tree</em>
the resulting structural tree with its support values. 
It comes along with the option to modify the tree using an external application. 
The user can also download the Newick format tree file (available in the <i>Result files</i> section) to use other visualization tools</li>

<li><em>Template list</em>
the list of template structure files associated with each sequence. 
The structure files are named after the PDB code and the identifier chain (ChainID) of the chain use for the computation. 
Structures are linked to their corresponding PDB webpage for more information.
</li>

<li><em>Result files</em>
the result files coming from the alignment and T-RMSD method; all files can be downloaded as a single zip file
</li>

<li><em>Citation</em>
the related article when citing the T-RMSD.
</li>

<li><em>Info</em>
some information related to the submitted job (running time, date, etc).
</li>

<li><em>Command line</em>
the command lines (separated by a semicolon) used by the webserver to run successively the Expresso alignment and the T-RMSD method. 
Lines can then be reused and/or modified to run locally with your own T-Coffee package.
</li>

<li><em>Replay</em>
this link allows the user to re-run the job while modifying input options or data.
</li>

</ul>

<p> <img alt="T-Coffee home" src="public/images/trmsd2.jpg" width="728" > </p>

</div>


<div class="box">
<h2><span>References</span></h2>

<div id="ref">
<ol style="list-style: decimal; padding-left: 1.4em">
<li>
	<a href='http://www.ncbi.nlm.nih.gov/pubmed/20471393' target='_blank' >
	T-RMSD: a fine-grained, structure-based classification method and its application to the functional characterization of TNF receptors</a>
	<i>Magis C, Stricher F, van der Sloot AM, Serrano L, Notredame C.</i>
</li>

<li>
	<a href='http://www.ncbi.nlm.nih.gov/pubmed/22789664' target='_blank' >
	An improved understanding of TNFL/TNFR interactions using structure-based classifications.
	</a>
	<i>Magis C, van der Sloot AM, Serrano L, Notredame C.</i>

</ol>
</div>

</div>
